### PR TITLE
Add DataType.String (DT_STRING) support

### PR DESCRIPTION
**Pull request by thomas-stripe**
This PR adds support for the DT_STRING ONNX data type. This is used to represent arbitrary byte strings. The internal type used for elements is `ByteString` from Protobuf. This was complicated because we had assumed all types had fixed width serialization previously, whereas string types are variable. There are 2 changes to address this:
 - Addition of a "strategy" to the `ToBytes` type class to determine whether a type supports a fixed or variable width encoding.
 - Addition of a new serialization format for variable length types in `Tensor`. In this case, we store an index of offsets (`int64`) for the variable length data first, followed by the variable length data all encoded sequentially. For byte strings, we encode them by first encoding the length as a var int, followed by the data as-is.

This PR does not currently add support for memory-mapped string tensors. This seems like a good thing to support, since string-based tensors can easily grow quite large, but we can do it in a separate PR.

### CODE REVIEW

File: core/src/main/scala/com/stripe/agate/tensor/OnnxNumber.scala [hunk 1]
```diff
@@ -525,7 +525,10 @@ object OnnxNumber {
     }
 
   def forDataType(dt: DataType): OnnxNumber[dt.Elem] =
-    forDataTypeMap(dt)
+    forDataTypeMap.get(dt: DataType.Aux[dt.Elem]) match {
+      case Some(num) => num
+      case None      => throw new IllegalArgumentException(s"not a numeric datatype=$dt")
```
Comment (thomas-stripe) [inline core/src/main/scala/com/stripe/agate/tensor/OnnxNumber.scala]: Just chucking an exception requires the least amount of change. We could update our numeric operations on tensor to return `Try`s instead, but figured I'd see how y'all feel about it.

File: core/src/main/scala/com/stripe/agate/tensor/Storage.scala [hunk 2]
```diff
@@ -212,49 +291,57 @@ object Storage {
...
-    val step = tb.size
-    if (maxChunkSizeInBytes < step) {
-      Resource.liftF(
-        IO.raiseError(
-          new IllegalArgumentException(s"maxChunkSizeInBytes=${maxChunkSizeInBytes} < $step")
+    tb.strategy match {
+      case _: ToBytes.Strategy.VarLength[_] =>
+        Resource.liftF(
+          IO.raiseError(
+            new IllegalArgumentException(s"unsupported datatype: $dt")
```
Comment (thomas-stripe) [inline core/src/main/scala/com/stripe/agate/tensor/Storage.scala]: Currently just chucking an exception. We'll need a different strategy for variable-length encoded values instead of using a `ChunkedStorage`. I'm going to do this in a separate PR though. We already throw for anything except `Float32` anyways, so I figure it's OK to throw for strings for the near term too.

**Comment by thomas-stripe** [comment]
OK, I think this is good to go. I added the string type so it gets generated for tests and they all pass now. I also added 1 test for the ToBytes implementation.

File: core/src/main/scala/com/stripe/agate/tensor/OnnxShow.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  val showBFloat16: Show[Short] = Show.show(n => new BFloat16(n).toString)
+  val showFloat16: Show[Short] = Show.show(n => new Float16(n).toString)
+  val showFloat32: Show[Float] = Show.fromToString
+  val showFloat64: Show[Double] = Show.fromToString
+  val showString: Show[ByteString] = Show.show { s =>
+    // Renders a Python-style bytes string.
+    val bldr = new StringBuilder("'")
+    val it = s.iterator()
+    while (it.hasNext) {
+      val c = it.nextByte().toInt
```
Comment (tixxit) [inline core/src/main/scala/com/stripe/agate/tensor/OnnxShow.scala]: This is a `ByteString.ByteIterator` and does not box here.

File: core/src/main/scala/com/stripe/agate/tensor/Tensor.scala [hunk 4]
```diff
@@ -112,19 +114,22 @@ abstract class Tensor[D <: DataType] {
...
+      case Success(num) =>
+        val s0 = axes
+        val s1 = that.axes
+        s0 == s1 && s0.forall { cs =>
+          val x0 = this(cs)
+          val x1 = that(cs)
+          Util.closeTo(num.toFloat(x0), num.toFloat(x1), error)
+        }
+      case _ =>
+        this == that
```
Comment (story-stripe) [inline core/src/main/scala/com/stripe/agate/tensor/Tensor.scala]: If I am reading this correctly, for all non-numeric data types we are checking strict equality here? 

If that is a case, can you add a small comment above: 
```suggestion
       // For all non-numeric datatypes, perform a strict equality check
        this == that
```

File: core/src/main/scala/com/stripe/agate/tensor/OnnxShow.scala [hunk 5]
```diff
@@ -0,0 +1,51 @@
+package com.stripe.agate.tensor
+
+import cats.Show
+import com.google.protobuf.ByteString
+import com.stripe.dagon.HMap
+
+object OnnxShow {
```
Comment (story-stripe) [inline core/src/main/scala/com/stripe/agate/tensor/OnnxShow.scala]: :+1: nice helper class

File: core/src/main/scala/com/stripe/agate/tensor/OnnxNumber.scala [hunk 6]
```diff
@@ -541,14 +534,20 @@ object OnnxNumber {
...
 
-  def cast(from: DataType, to: DataType): from.Elem => to.Elem = {
-    val onF = forDataType(from)
-    val onT = forDataType(to)
-
-    if (onT.isInstanceOf[OnnxIntegral[_]]) { (a: from.Elem) =>
-      onT.fromLong(onF.toLong(a))
-    } else { (a: from.Elem) =>
-      onT.fromDouble(onF.toDouble(a))
+  def cast(from: DataType, to: DataType): Option[from.Elem => to.Elem] =
```
Comment (story-stripe) [inline core/src/main/scala/com/stripe/agate/tensor/OnnxNumber.scala]: This feels a little more like a `Try` than an `Option` here to me, but I don't feel super strongly about it.

File: core/src/main/scala/com/stripe/agate/tensor/OnnxShow.scala [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  val showFloat16: Show[Short] = Show.show(n => new Float16(n).toString)
+  val showFloat32: Show[Float] = Show.fromToString
+  val showFloat64: Show[Double] = Show.fromToString
+  val showString: Show[ByteString] = Show.show { s =>
+    // Renders a Python-style bytes string.
+    val bldr = new StringBuilder("'")
+    val it = s.iterator()
+    while (it.hasNext) {
+      val c = it.nextByte().toInt
+      if ((c >= 32 && c <= 126) && (c != '\''.toInt) && (c != '\\'.toInt)) {
```
[the remark this answers is not in the document]
Comment (tixxit) [reply core/src/main/scala/com/stripe/agate/tensor/OnnxShow.scala]: By default, there isn't a ton. An alternative is to base64 encode it, which is supported out of the box. Another alternative would be to try and render it as a string that Scala/Java can parse, using the `\uXXXX` syntax, but that just seemed weird, given that we're rendering bytes, not unicode code points, so only ever use the last 2 digits. I thought the Python bytes syntax was kind of nice, so went with that, but happy to switch too!

File: core/src/main/scala/com/stripe/agate/tensor/OnnxNumber.scala (hunk 6 above)
Comment (tixxit) [reply core/src/main/scala/com/stripe/agate/tensor/OnnxNumber.scala]: Yeah, I thought about it a bit too, but in all the places we're using it, we are just treating it as an option. Like: "give me this function if it exists, otherwise I'll do something else".

**Comment by tixxit** [comment]
@johnynek no worries and thanks! Last week was pretty quiet in general :)
